File: app/src/test/java/com/rifafauzi/moviecatalogue/helper/repository/RepositoryTest.java
package com.rifafauzi.moviecatalogue.helper.repository;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.MutableLiveData;

import com.rifafauzi.moviecatalogue.helper.local.LocalDataSource;
import com.rifafauzi.moviecatalogue.helper.local.entity.MoviesEntity;
import com.rifafauzi.moviecatalogue.helper.local.entity.TvShowEntity;
import com.rifafauzi.moviecatalogue.helper.remote.RemoteDataSource;
import com.rifafauzi.moviecatalogue.helper.vo.Resource;
import com.rifafauzi.moviecatalogue.model.Movies;
import com.rifafauzi.moviecatalogue.utils.FakeData;
import com.rifafauzi.moviecatalogue.utils.InstantAppExecutors;
import com.rifafauzi.moviecatalogue.utils.LiveDataTestUtil;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import java.util.List;

import static org.junit.Assert.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class RepositoryTest {

    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    private LocalDataSource local = mock(LocalDataSource.class);
    private RemoteDataSource remote = mock(RemoteDataSource.class);
    private InstantAppExecutors instantAppExecutors = mock(InstantAppExecutors.class);
    private FakeRepository fakeRepository = new FakeRepository(remote, local, instantAppExecutors);

    private List<Movies> moviesResponses = FakeData.generateDummyApiMovies();
    private List<MoviesEntity> moviesEntities = FakeData.generateDummyLocalMovies();
    private int moviesId = moviesResponses.get(0).getId();

    private List<Movies> tvShowResponses = FakeData.generateDummyApiTvShow();
    private List<TvShowEntity> tvShowEntities = FakeData.generateDummyLocalTvShow();
    private int tvShowId = tvShowResponses.get(0).getId();

    @Before
    public void setUp() {

    }

    @After
    public void tearDown() {

    }

    @Test
    public void getAllMovies() {
        MutableLiveData<List<MoviesEntity>> listMutableLiveData = new MutableLiveData<>();
        listMutableLiveData.setValue(moviesEntities);

        when(local.getAllMovies()).thenReturn(listMutableLiveData);

        Resource<List<MoviesEntity>> result = LiveDataTestUtil.getValue(fakeRepository.getAllMovies());

        verify(local).getAllMovies();
        assertNotNull(result.data);
        assertEquals(moviesResponses.size(), result.data.size());
    }

    @Test
    public void getDetailMovies() {
        MutableLiveData<MoviesEntity> dummyEntity = new MutableLiveData<>();
        dummyEntity.setValue(FakeData.getMovieDetail(FakeData.generateDummyLocalMovies().get(0), false));

        when(local.getDetailMovies(moviesId)).thenReturn(dummyEntity);

        Resource<MoviesEntity> result = LiveDataTestUtil.getValue(fakeRepository.getDetailMovies(moviesId));

        verify(local).getDetailMovies(moviesId);
        assertNotNull(result.data);
        assertNotNull(result.data.getTitle());
        assertEquals(moviesResponses.get(0).getTitle(), result.data.getTitle());
    }



    @Test
    public void getAllTvShow() {
        MutableLiveData<List<TvShowEntity>> listMutableLiveData = new MutableLiveData<>();
        listMutableLiveData.setValue(tvShowEntities);

        when(local.getAllTvShow()).thenReturn(listMutableLiveData);

        Resource<List<TvShowEntity>> result = LiveDataTestUtil.getValue(fakeRepository.getAllTvShow());

        verify(local).getAllTvShow();
        assertNotNull(result.data);
        assertEquals(tvShowResponses.size(), result.data.size());
    }

    @Test
    public void getDetailTvShow() {
        MutableLiveData<TvShowEntity> dummyEntity = new MutableLiveData<>();
        dummyEntity.setValue(FakeData.getTvShowsDetail(FakeData.generateDummyLocalTvShow().get(0), false));

        when(local.getDetailTvShow(tvShowId)).thenReturn(dummyEntity);

        Resource<TvShowEntity> result = LiveDataTestUtil.getValue(fakeRepository.getDetailTvShow(tvShowId));

        verify(local).getDetailTvShow(tvShowId);
        assertNotNull(result.data);
        assertNotNull(result.data.getName());
        assertEquals(tvShowResponses.get(0).getName(), result.data.getName());
    }

}

File: app/src/test/java/com/rifafauzi/moviecatalogue/helper/repository/RepositoryTest.kt
package com.rifafauzi.moviecatalogue.helper.repository

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import com.rifafauzi.moviecatalogue.helper.remote.RemoteDataSource
import com.rifafauzi.moviecatalogue.model.Movies
import com.rifafauzi.moviecatalogue.model.TvShow
import com.rifafauzi.moviecatalogue.utils.FakeData
import com.rifafauzi.moviecatalogue.utils.LiveDataTestUtil
import org.junit.*
import org.mockito.Mockito

class RepositoryTest {

    @Rule
    @JvmField
    var instantTaskExecutorRule = InstantTaskExecutorRule()

    private val remoteRepository = Mockito.mock(RemoteDataSource::class.java)
    private val dataRepositoryTest = FakeRepository(remoteRepository)

    private val movieList = FakeData.generateDummyMovies()
    private val movieDetail = FakeData.getMovieDetail()
    private val movieId = movieList[0].id.toString()
    private lateinit var movie : Movies

    private val tvShowsList = FakeData.generateDummyTvShow()
    private val tvShowsDetail = FakeData.getTvShowsDetail()
    private val tvShowId = tvShowsList[0].id.toString()
    private lateinit var tvShow : TvShow



    @Before
    fun setUp() {
        movie = Movies(
                "/dzBtMocZuJbjLOXvrl4zGYigDzh.jpg",
                "Simba idolises his father, King Mufasa, and takes to heart his own royal destiny. But not everyone in the kingdom celebrates the new cub's arrival. Scar, Mufasa's brother—and former heir to the throne—has plans of his own. The battle for Pride Rock is ravaged with betrayal, tragedy and drama, ultimately resulting in Simba's exile. With help from a curious pair of newfound friends, Simba will have to figure out how to grow up and take back what is rightfully his.",
                "2019-07-12",
                420818,
                "The Lion King"
        )

        tvShow = TvShow(
                "/b71BaRjp9TwxUZodLGgSRIlkfL8.jpg",
                "The dramatisation of one of the most notorious killing sprees in British history.",
                "",
                11634,
                ""
        )
    }

    @After
    fun tearDown() {

    }

    @Test
    fun getTvShowsDetail() {
        Mockito.doAnswer {
            val callback = it.arguments[1] as RemoteDataSource.GetTvShowDetailCallback
            callback.onTvShowDetailLoaded(tvShowsDetail)
            null
        }.`when`(remoteRepository).getTvShowDetail(
                Mockito.eq(tvShowId),
                Mockito.any(RemoteDataSource.GetTvShowDetailCallback::class.java))

        val result = LiveDataTestUtil.getValue(dataRepositoryTest.getTvShowsDetail(tvShowId))
        Mockito.verify(
                remoteRepository,
                Mockito.times(1)).getTvShowDetail(Mockito.eq(tvShowId), Mockito.any(RemoteDataSource.GetTvShowDetailCallback::class.java))
        Assert.assertEquals(tvShowsDetail.id, result.id)

    }

    @Test
    fun getTvShowsList() {
        Mockito.doAnswer {
            val callback = it.arguments[0] as RemoteDataSource.GetTvShowCallback
            callback.onTvShowLoaded(tvShowsList)
            null
        }.`when`(remoteRepository).getListTvShow(Mockito.any(RemoteDataSource.GetTvShowCallback::class.java))

        val result = LiveDataTestUtil.getValue(dataRepositoryTest.getTvShowsList())
        Mockito.verify(remoteRepository, Mockito.times(1)).getListTvShow(Mockito.any(RemoteDataSource.GetTvShowCallback::class.java))

        Assert.assertEquals(tvShowsList.size, result.size)
    }

    @Test
    fun getMovieDetail() {
        Mockito.doAnswer {
            val callback = it.arguments[1] as RemoteDataSource.GetMoviesDetailCallback
            callback.onMoviesDetailLoaded(movieDetail)
            null
        }.`when`(remoteRepository).getMoviesDetail(Mockito.eq(movieId), Mockito.any(RemoteDataSource.GetMoviesDetailCallback::class.java))

        val result = LiveDataTestUtil.getValue(dataRepositoryTest.getMovieDetail(movieId))
        Mockito.verify(
                remoteRepository,
                Mockito.times(1)).getMoviesDetail(Mockito.eq(movieId), Mockito.any(RemoteDataSource.GetMoviesDetailCallback::class.java))
        Assert.assertEquals(movieDetail.id, result.id)
    }

    @Test
    fun getMovie() {
        Mockito.doAnswer {
            val callback = it.arguments[0] as RemoteDataSource.GetMoviesCallback
            callback.onMoviesLoaded(movieList)
            null
        }.`when`(remoteRepository).getListMovies(Mockito.any(RemoteDataSource.GetMoviesCallback::class.java))

        val result = LiveDataTestUtil.getValue(dataRepositoryTest.getMovieList())
        Mockito.verify(remoteRepository, Mockito.times(1)).getListMovies(Mockito.any(RemoteDataSource.GetMoviesCallback::class.java))

        Assert.assertEquals(movieList.size, result.size)
    }

}